File: android/app/src/main/java/com/opencontinuity/core/session/SessionManager.kt
package com.opencontinuity.core.session

import android.util.Log
import com.opencontinuity.core.connection.ConnectionManager
import com.opencontinuity.core.protocol.MessageType
import kotlinx.coroutines.*
import java.util.concurrent.ConcurrentHashMap
import javax.crypto.SecretKey

class SessionManager(private val connectionManager: ConnectionManager) {

    companion object {
        private const val TAG = "SessionManager"
        private const val HEARTBEAT_TIMEOUT_MS = 30_000L
        private const val WATCHDOG_INTERVAL_MS = 5_000L
    }

    data class DeviceSession(
        val deviceId: String,
        var deviceName: String,
        var sessionToken: String,
        var encryptionKey: SecretKey? = null,
        val capabilities: List<String> = emptyList(),
        val connectedAt: Long = System.currentTimeMillis(),
        @Volatile var lastHeartbeat: Long = System.currentTimeMillis(),
        @Volatile var reconnectCount: Int = 0
    )

    private val sessions = ConcurrentHashMap<String, DeviceSession>()
    private var scope: CoroutineScope? = null
    private var watchdogJob: Job? = null

    var onSessionExpired: ((DeviceSession) -> Unit)? = null
    var onSessionStarted: ((DeviceSession) -> Unit)? = null
    var onSessionRestored: ((DeviceSession) -> Unit)? = null

    fun start() {
        scope?.cancel()
        scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

        connectionManager.registerHandler(MessageType.HEARTBEAT) { sessionId, _ ->
            touchHeartbeatForWsSession(sessionId)
        }

        connectionManager.registerHandler(MessageType.HEARTBEAT_ACK) { sessionId, _ ->
            touchHeartbeatForWsSession(sessionId)
        }

        startWatchdog()
    }

    fun stop() {
        watchdogJob?.cancel()
        watchdogJob = null
        scope?.cancel()
        scope = null
    }

    fun createSession(
        deviceId: String,
        deviceName: String,
        sessionToken: String,
        capabilities: List<String> = emptyList(),
        encryptionKey: SecretKey? = null
    ): DeviceSession {
        val session = DeviceSession(
            deviceId = deviceId,
            deviceName = deviceName,
            sessionToken = sessionToken,
            encryptionKey = encryptionKey,
            capabilities = capabilities
        )
        sessions[deviceId] = session
        onSessionStarted?.invoke(session)
        Log.i(TAG, "Session started: $deviceName ($deviceId)")
        return session
    }

    fun restoreSession(deviceId: String, deviceName: String, newToken: String): DeviceSession? {
        val existing = sessions[deviceId] ?: return null
        existing.deviceName = deviceName
        existing.sessionToken = newToken
        existing.lastHeartbeat = System.currentTimeMillis()
        existing.reconnectCount++
        onSessionRestored?.invoke(existing)
        Log.i(TAG, "Session restored: $deviceName (reconnect #${existing.reconnectCount})")
        return existing
    }

    fun getSession(deviceId: String): DeviceSession? = sessions[deviceId]
    fun getAllSessions(): List<DeviceSession> = sessions.values.toList()

    fun setEncryptionKey(deviceId: String, key: SecretKey) {
        sessions[deviceId]?.encryptionKey = key
    }

    fun endSession(deviceId: String) {
        sessions.remove(deviceId)?.let { Log.i(TAG, "Session ended: ${it.deviceName}") }
    }

    fun touchHeartbeat(deviceId: String) {
        sessions[deviceId]?.lastHeartbeat = System.currentTimeMillis()
    }

    private fun touchHeartbeatForWsSession(wsSessionId: String) {
        val clientName = connectionManager.connectedClients.value
            .firstOrNull { it.sessionId == wsSessionId }?.deviceName ?: return
        sessions.values.firstOrNull { it.deviceName == clientName }
            ?.let { touchHeartbeat(it.deviceId) }
    }

    private fun startWatchdog() {
        if (watchdogJob?.isActive == true) return
        watchdogJob = scope?.launch {
            while (isActive) {
                delay(WATCHDOG_INTERVAL_MS)
                checkHeartbeats()
            }
        }
    }

    private fun checkHeartbeats() {
        val now = System.currentTimeMillis()
        for (session in sessions.values.toList()) {
            val elapsed = now - session.lastHeartbeat
            if (elapsed > HEARTBEAT_TIMEOUT_MS) {
                Log.w(TAG, "Heartbeat timeout for ${session.deviceName} (${elapsed}ms)")
                sessions.remove(session.deviceId)
                onSessionExpired?.invoke(session)
            }
        }
    }
}

File: windows/src/main/session/SessionManager.ts
/**
 * Session Manager
 *
 * Maintains the active device session across reconnects.
 * Responsibilities:
 *   - One session per paired device
 *   - 5-second heartbeat ping, 10-second timeout → reconnect
 *   - Stores per-session encryption key derived during pairing
 *   - Emits 'sessionStarted', 'sessionRestored', 'sessionExpired' events
 */

import { EventEmitter } from 'events';
import crypto from 'crypto';

export interface DeviceSession {
  /** Stable UUID assigned at first pairing */
  deviceId: string;
  deviceName: string;
  /** Runtime WebSocket session token (changes on reconnect) */
  sessionToken: string;
  /** AES-256 key derived via ECDH + HKDF — persists across reconnects */
  encryptionKey?: Buffer;
  capabilities: string[];
  connectedAt: number;
  /** Updated on every heartbeat_ack */
  lastHeartbeat: number;
  reconnectCount: number;
}

export class SessionManager extends EventEmitter {
  /** One session per deviceId */
  private sessions = new Map<string, DeviceSession>();

  /** Watchdog checks heartbeats every 5 s */
  private readonly HEARTBEAT_INTERVAL_MS = 5_000;
  /** Session is considered dead if no heartbeat within 10 s */
  private readonly HEARTBEAT_TIMEOUT_MS = 10_000;

  private watchdog: NodeJS.Timeout | null = null;

  // ─────────────────────────────────────────── lifecycle ──────────────────

  start(): void {
    if (this.watchdog) return;
    this.watchdog = setInterval(() => this.checkHeartbeats(), this.HEARTBEAT_INTERVAL_MS);
  }

  stop(): void {
    if (this.watchdog) {
      clearInterval(this.watchdog);
      this.watchdog = null;
    }
  }

  // ───────────────────────────────────────── session CRUD ─────────────────

  /**
   * Create a brand-new session when a device connects for the first time.
   */
  createSession(
    deviceId: string,
    deviceName: string,
    capabilities: string[],
    encryptionKey?: Buffer
  ): DeviceSession {
    const session: DeviceSession = {
      deviceId,
      deviceName,
      sessionToken: crypto.randomBytes(32).toString('base64'),
      encryptionKey,
      capabilities,
      connectedAt: Date.now(),
      lastHeartbeat: Date.now(),
      reconnectCount: 0
    };
    this.sessions.set(deviceId, session);
    this.emit('sessionStarted', session);
    console.log(`[SessionManager] Session started: ${deviceName} (${deviceId})`);
    return session;
  }

  /**
   * Restore an existing session when a previously-paired device reconnects.
   * Preserves the encryption key so the connection is immediately encrypted.
   */
  restoreSession(deviceId: string, deviceName: string): DeviceSession | null {
    const existing = this.sessions.get(deviceId);
    if (!existing) return null;

    existing.deviceName    = deviceName;
    existing.sessionToken  = crypto.randomBytes(32).toString('base64');
    existing.lastHeartbeat = Date.now();
    existing.reconnectCount++;

    this.emit('sessionRestored', existing);
    console.log(`[SessionManager] Session restored: ${deviceName} (reconnect #${existing.reconnectCount})`);
    return existing;
  }

  getSession(deviceId: string): DeviceSession | undefined {
    return this.sessions.get(deviceId);
  }

  getSessionByToken(token: string): DeviceSession | undefined {
    for (const s of this.sessions.values()) {
      if (s.sessionToken === token) return s;
    }
    return undefined;
  }

  getAllSessions(): DeviceSession[] {
    return Array.from(this.sessions.values());
  }

  setEncryptionKey(deviceId: string, key: Buffer): void {
    const s = this.sessions.get(deviceId);
    if (s) s.encryptionKey = key;
  }

  endSession(deviceId: string): void {
    const s = this.sessions.get(deviceId);
    if (s) {
      this.sessions.delete(deviceId);
      this.emit('sessionEnded', s);
      console.log(`[SessionManager] Session ended: ${s.deviceName}`);
    }
  }

  // ─────────────────────────────────────── heartbeat tracking ─────────────

  touchHeartbeat(deviceId: string): void {
    const s = this.sessions.get(deviceId);
    if (s) s.lastHeartbeat = Date.now();
  }

  private checkHeartbeats(): void {
    const now = Date.now();
    for (const [deviceId, session] of this.sessions) {
      const elapsed = now - session.lastHeartbeat;
      if (elapsed > this.HEARTBEAT_TIMEOUT_MS) {
        console.warn(
          `[SessionManager] Heartbeat timeout for ${session.deviceName} (${elapsed}ms) — emitting sessionExpired`
        );
        this.emit('sessionExpired', session);
        // Don't remove: ConnectionManager will call endSession after it attempts reconnect
      }
    }
  }
}
